src/miniso.rs

```diff
@@ -0,0 +1,271 @@
+            version: HEADER_VERSION,
+            app_version: crate_version!().into(),
+        }
+    }
+}
+
+impl Header {
+    pub fn validate(&self) -> Result<()> {
+        if self.magic != HEADER_MAGIC {
+            bail!("not a miniso file!");
+        }
+        if self.version != HEADER_VERSION {
+            bail!(
+                "incompatible miniso file version: {} vs {} (created by {})",
+                HEADER_VERSION,
+                self.version,
+                self.app_version,
+            );
+        }
+        Ok(())
+    }
+}
+
+// Version-specific payload. Evolvable.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Data {
+    table: Table,
+    digest: Sha256Digest,
+    xzpacked: Vec<u8>,
+}
+
+impl Data {
+    pub fn xzpack(
+        miniso: &mut File,
+        full_files: &HashMap<String, iso9660::File>,
+        minimal_files: &HashMap<String, iso9660::File>,
+    ) -> Result<(Self, usize, u64, u64, u64)> {
+        let table = Table::new(full_files, minimal_files)?;
+
+        // A `ReadHasher` here would let us wrap the miniso so we calculate the digest as we read.
+        let digest = Sha256Digest::from_file(miniso)?;
+        let mut offset = miniso
+            .seek(SeekFrom::Start(0))
+            .context("seeking back to miniso start")?;
+
+        let mut xzw = XzEncoder::new(Vec::new(), 9);
+        let mut buf = [0u8; BUFFER_SIZE];
+        let mut skipped: u64 = 0;
+        for entry in &table.entries {
+            let addr: u64 = entry.minimal.as_offset();
+            assert!(offset <= addr);
+            if addr > offset {
+                copy_exactly_n(miniso, &mut xzw, addr - offset, &mut buf).with_context(|| {
+                    format!(
+                        "copying {} miniso bytes at offset {}",
+                        addr - offset,
+                        offset
+                    )
+                })?;
+            }
+            // I tested trying to be smarter here and rounding to the nearest 2k block so we can
+            // skip padding, but zeroes compress so well that it only saved a grand total of 4
+            // bytes after xz. So not worth the complexity.
+            offset = miniso
+                .seek(SeekFrom::Current(entry.length as i64))
+                .with_context(|| format!("skipping miniso file at offset {}", addr))?;
+            skipped += entry.length as u64;
+        }
+
+        copy(miniso, &mut xzw).context("copying remaining miniso bytes")?;
+
+        xzw.try_finish().context("trying to finish xz stream")?;
+        let matches = table.entries.len();
+        let written = xzw.total_in();
+        let written_compressed = xzw.total_out();
+        Ok((
+            Self {
+                table,
+                digest,
+                xzpacked: xzw.finish().context("finishing xz stream")?,
+            },
+            matches,
+            skipped,
+            written,
+            written_compressed,
```

Comment:
Was really tempted to return a struct like `XzPackStats` or something here to group these loose return values. But... let's just keep that as a follow-up.